File: perl/keyword-SERPs.pl
#!/usr/bin/perl -w
#
# GitHub: https://github.com/SERPWoo/API
#
# This code requests a Keyword's SERP Results and outputs the Timestamp, Rank, Type, Page Title, URL
#
# This output is text format
#
# Last updated - Aug 24th, 2017 @ 16:52 EST (@MercenaryCarter https://github.com/MercenaryCarter and https://twitter.com/MercenaryCarter)
#
# Run Command: perl list-SERPs-for-keyword.pl

	use strict;
	use warnings;
	
	use LWP::UserAgent; #You may need to install LWP::Protocol::https through cpan
	use JSON qw( decode_json );

	use utf8;
	use open ':std', ':encoding(UTF-8)';

	#outputs data in JSON format

		# Get your API Key here: https://www.serpwoo.com/v3/api/ (should be logged in)
			my $API_key = "API_KEY_HERE";

			my $Project_ID = PROJECT_ID_HERE; #Integer, example = 12345
			my $Keyword_ID = KEYWORD_ID_HERE; #Integer, example = 456789

			my %JSON_DATA;
	
		#Pulls the S3 JSON file
			my $ua = new LWP::UserAgent;
			my $res = $ua->get('https://api.serpwoo.com/v1/serps/' . $Project_ID . '/' . $Keyword_ID . '/?key=' . $API_key);

			my $json_content = $res->decoded_content;
			
			my $error_occurred = 0;
			my $rank;
			my $type;
			my $title;
			my $url;

				eval { %JSON_DATA = %{ decode_json($json_content) }; } or $error_occurred = 1;
				
				if ($error_occurred == 0) {

					# Always check for success = 1
					if ($JSON_DATA{'success'} == 1){

						print "\n--\n";
					    printf "%-15s %-10s %-10s %-80s %-80s\n", "Timestamp", "Rank", "Type", "Title", "URL";
					    printf "%-15s %-10s %-10s %-80s %-80s\n", "---------", "----", "----", "-----", "---";

							foreach my $timestamp (sort {$b <=> $a} keys %{$JSON_DATA{$Keyword_ID}}) {

							    #printf "Timestamp: $timestamp\n\n";

									foreach my $rank (sort {$a <=> $b} keys %{$JSON_DATA{$Keyword_ID}{$timestamp}{'results'}}) {
										$type = $JSON_DATA{$Keyword_ID}{$timestamp}{'results'}{$rank}{'type'};
										$title = $JSON_DATA{$Keyword_ID}{$timestamp}{'results'}{$rank}{'title'};
										$url = $JSON_DATA{$Keyword_ID}{$timestamp}{'results'}{$rank}{'url'};
								
										    printf "%-15s %-10s %-10s %-80s %-80s\n", $timestamp, $rank, $type, $title, $url;
									}								

								    print "--\n";

							}


					}else {
						#Something went wrong, outputs message
						print "Problem. Error: " . $JSON_DATA{'error'} . "\n";
					}
					
				}else {
						print "Something went VERY wrong.\n\n";
						#print "$json_content";
				}

	  		  print "\n--\n";



################################################################################
################################################################################

File: python/keyword-SERPs.py
#
# GitHub: https://github.com/SERPWoo/API
#
# This code requests a Keyword's SERP Results and outputs the Timestamp, Rank, Type, Page Title, URL
#
# This output is text format
#
# Last updated - Aug 24th, 2017 @ 23:05 EST (@MercenaryCarter https://github.com/MercenaryCarter and https://twitter.com/MercenaryCarter)
#
# Run Command: python keyword-SERPs.py

import requests, json

def main():        
    API_key = 'API_KEY_HERE'
    Project_ID = 0 #input your Project ID
    Keyword_ID = 0 #input your Keyword ID

    url = "https://api.serpwoo.com/v1/serps/" + str(Project_ID) + "/" + str(Keyword_ID) + "/"
    payload = {'key': API_key}

    r = requests.get(url, params=payload)
    j = r.json()

    print('\n--\n')

    print('%-15s %-10s %-10s %-80s %-80s' % ("Timestamp", "Rank", "Type", "Title", "URL"))
    print('%-15s %-10s %-10s %-80s %-80s' % ("---------", "----", "----", "-----", "---"))

    if j['success'] == 1:
        for keyword_id in j:
            
            #print('keyword_id=[' + keyword_id + '][' + str(Keyword_ID) + ']')            
            if keyword_id == str(Keyword_ID):
                #print('YES=[' + keyword_id + '][' + str(Keyword_ID) + ']')

                for timestamp in sorted(j[keyword_id], cmp=_compare_keys):
                    for rank in sorted(j[keyword_id][timestamp]['results'], cmp=_compare_keys):
                        print('%-15s %-10s %-10s %-80s %-80s' % (timestamp, rank, j[keyword_id][timestamp]['results'][rank]['type'], j[keyword_id][timestamp]['results'][rank]['title'], j[keyword_id][timestamp]['results'][rank]['url']))
    else:

        print('An Error Occurred: ' + j['error'])

    print('\n--\n')

################################################################################
########################################
########################################

#Sorting comparion
def _compare_keys(x, y):
    try:
        x = int(x)
    except ValueError:
        xint = False
    else:
        xint = True
    try:
        y = int(y)
    except ValueError:
        if xint:
            return -1
        return cmp(x.lower(), y.lower())
        # or cmp(x, y) if you want case sensitivity.
    else:
        if xint:
            return cmp(x, y)
        return 1


main()

################################################################################
################################################################################